src/DotVVM.Framework/Compilation/Static/StaticViewCompiler.cs

```diff
@@ -0,0 +1,277 @@
+        //     into one assembly. It's horrible, I know, but the compiler is riddled with references to System.Type
+        //     that make it presently impossible to compile it all in one go.
+        private readonly ConcurrentDictionary<string, StaticView> viewCache
+            = new ConcurrentDictionary<string, StaticView>();
+
+        public StaticViewCompiler(DotvvmConfiguration configuration, Assembly dotvvmProjectAssembly)
+        {
+            this.configuration = configuration;
+            this.dotvvmProjectAssembly = dotvvmProjectAssembly;
+            baseReferences = GetBaseReferences(configuration);
+        }
+
+        public static DotvvmConfiguration CreateConfiguration(
+            Assembly dotvvmProjectAssembly,
+            string dotvvmProjectDir)
+        {
+            InitializeDotvvmControls(dotvvmProjectAssembly);
+            return ConfigurationInitializer.GetConfiguration(dotvvmProjectAssembly, dotvvmProjectDir, services =>
+            {
+                services.AddSingleton<IControlResolver, StaticViewControlResolver>();
+                services.TryAddSingleton<IViewModelProtector, FakeViewModelProtector>();
+                services.AddSingleton(new RefObjectSerializer());
+                // Yes, this is here so that there can be a circular dependency in StaticViewControlResolver.
+                // I'm not happy about it, no, but the alternative is a more-or-less complete rewrite.
+                services.AddSingleton(p => new StaticViewCompiler(
+                    p.GetRequiredService<DotvvmConfiguration>(),
+                    dotvvmProjectAssembly));
+
+                // HACK: IDotvvmCacheAdapter is not in v2.0.0 that's why it's hacked this way.
+                var iCacheAdapter = Type.GetType(IDotvvmCacheAdapterName);
+                if (iCacheAdapter is object)
+                {
+                    services.AddSingleton(iCacheAdapter, Type.GetType(SimpleDictionaryCacheAdapterName));
+                }
+
+                // TODO: Uncomment when the views can actually be compiled into one assembly.
+                // var bindingCompiler = new AssemblyBindingCompiler(
+                //     assemblyName: null,
+                //     className: null,
+                //     outputFileName: null,
+                //     configuration: null);
+                // services.AddSingleton<IBindingCompiler>(bindingCompiler);
+                // services.AddSingleton<IExpressionToDelegateCompiler>(bindingCompiler.GetExpressionToDelegateCompiler());
+            });
+        }
+
+        public StaticView GetView(string viewPath)
+        {
+            if (viewCache.ContainsKey(viewPath))
+            {
+                return viewCache[viewPath];
+            }
+
+            var view = CompileView(viewPath);
+            // NB: in the meantime, the view could have been compiled on another thread
+            return viewCache.GetOrAdd(viewPath, view);
```

Comment:
Why not `viewCache.GetOrAdd(viewPath, CompileView)`?

Also, you might want to use `ConcurrentDictionary<..., Lazy<...>>` to prevent parallel compilation of one view (it is fairly expensive, so it seems worth it to me)

Reply:
Fixed by virtue of the `DefaultControlBuilderFactory` doing it this way.